assets/src/edit-story/components/fontPicker/pickerContainer.js

```diff
@@ -0,0 +1,304 @@
+  const listContainerRef = useRef();
+  const listRef = useRef();
+  const inputRef = useRef();
+  const currentActiveRef = useRef(0);
+
+  const [currentIndex, setCurrentIndex] = useState(0);
+  const [searchValue, setSearchValue] = useState('');
+  const [menuFonts, setMenuFonts] = useState([]);
+
+  // Recently used font options, clean up on refresh
+  const recentUsedFonts = useMemo(
+    () =>
+      searchValue && searchValue !== ''
+        ? recentUsedFontSlugs
+            .map((slug) => fonts.find((font) => font.slug === slug))
+            .filter(({ name }) =>
+              name.toLowerCase().includes(searchValue.toLowerCase())
+            )
+        : recentUsedFontSlugs.map((slug) =>
+            fonts.find((font) => font.slug === slug)
+          ),
+    [searchValue, recentUsedFontSlugs, fonts]
+  );
+
+  // Font options that start with search term
+  const normalFonts = useMemo(
+    () =>
+      searchValue && searchValue !== ''
+        ? fonts.filter(({ name }) =>
+            name.toLowerCase().startsWith(searchValue.toLowerCase())
+          )
+        : fonts,
+    [searchValue, fonts]
+  );
+
+  // Font options that include search term but not start with
+  const includeSearchFonts = useMemo(
+    () =>
+      searchValue && searchValue !== ''
+        ? fonts.filter(
+            ({ name }) =>
+              name.toLowerCase().indexOf(searchValue.toLowerCase()) > 0
+          )
+        : [],
+    [searchValue, fonts]
+  );
+
+  useEffect(() => {
+    const combinedFontList = []
+      .concat(recentUsedFonts)
+      .concat(normalFonts)
+      .concat(includeSearchFonts)
+      .slice(currentIndex, currentIndex + 10)
+      .map(({ name }) => name);
+    getMenuFonts(combinedFontList).then((result) => {
+      const resultArray = result.replace(/}/g, '}},').split('},');
+      setMenuFonts(resultArray);
+    });
+  }, [
+    currentIndex,
+    getMenuFonts,
+    recentUsedFonts,
+    normalFonts,
+    includeSearchFonts,
+  ]);
+
+  useEffect(() => {
+    inputRef.current.focus();
+  });
+
+  useEffect(() => {
+    const listElement = listContainerRef.current;
+
+    const handleScroll = () => {
+      let { scrollTop } = listElement;
+      scrollTop -= 5;
+      let currentVisibleIndex = Math.floor(scrollTop / 34);
+      if (
+        recentUsedFonts.length > 0 &&
+        currentVisibleIndex > recentUsedFonts.length
+      ) {
+        scrollTop -= 10;
+        currentVisibleIndex =
+          Math.floor(scrollTop / 34) - recentUsedFonts.length;
+      }
+      if (normalFonts.length > 0 && currentVisibleIndex > normalFonts.length) {
+        scrollTop -= 10;
+      }
+      currentVisibleIndex = Math.floor(scrollTop / 34);
+      if (currentVisibleIndex < 0) {
+        currentVisibleIndex = 0;
+      }
+      if (currentVisibleIndex !== currentActiveRef.current) {
+        currentActiveRef.current = currentVisibleIndex;
+        setCurrentIndex(currentVisibleIndex);
+      }
+    };
```

Comment:
Could you please document what each step of this logic is doing?

Reply:
Done